### ISSUE TITLE
Format.printf raises exception when using multiple domains

### ISSUE THREAD

**Issue by talex5** [ocamllabs/ocaml-multicore#563]
Running this code with `4.12.0+multicore` gives `Stdlib.Queue.Empty`:
```ocaml
let () =
  let domains = Array.init 7 (fun i ->
    Domain.spawn (fun () ->
          for j = 1 to 10000000 do () done;
          Format.printf "Running in domain %d@." i;
          flush stdout;
      )
    ) in
  Array.iter Domain.join domains
```

```
$ ocamlopt -o multi.exe multi.ml && ./multi.exe
Running in domain 0
Running in domain 
1
5
Running in domain Running in domain 2Fatal error: exception Stdlib.Queue.Empty
```

**Comment by kayceesrk** [comment]
The backtraces were not useful as the error was being raised by one domain, which is re-raised by the joining domain.

```bash
$ OCAMLRUNPARAM="b=1" ./a.out 
Running in domain 0
Running in domain 6

4
Running in domain 5
Fatal error: exception Stdlib.Queue.Empty
Raised at Stdlib__domain.join in file "domain.ml", line 113, characters 16-24
Called from Stdlib__array.iter in file "array.ml", line 95, characters 31-48
Called from Format_bug in file "format_bug.ml", line 9, characters 2-32
```

But `gdb` produces useful backtrace:

```bash
$ gdb ./a.out 
(gdb) break caml_raise_exn
Breakpoint 1 at 0x978c4
(gdb) r
Starting program: /home/kc/temp/a.out 
...
Thread 6 "a.out" hit Breakpoint 1, 0x00005555555eb8c4 in caml_raise_exn ()
(gdb) bt
#0  0x00005555555eb8c4 in caml_raise_exn ()
#1  0x00005555555ab7ef in camlStdlib__queue__take_117 () at queue.ml:73
#2  0x00005555555c3a04 in camlStdlib__format__advance_left_398 () at format.ml:436
#3  0x00005555555c7310 in camlStdlib__format__output_acc_1138 () at format.ml:1298
#4  0x00005555555c702b in camlStdlib__format__output_acc_1138 () at format.ml:1288
#5  0x00005555555c7823 in camlStdlib__format__fun_2403 () at format.ml:1349
#6  0x0000555555597fc4 in camlFormat_bug__fun_366 ()
#7  0x00005555555cb5a9 in camlStdlib__domain__body_208 () at domain.ml:71
#8  <signal handler called>
#9  0x00005555555e04f0 in caml_callback_exn (closure=<optimized out>, arg=<optimized out>, arg@entry=1) at callback.c:164
#10 0x00005555555e0a39 in caml_callback (closure=<optimized out>, arg=arg@entry=1) at callback.c:247
#11 0x00005555555e91be in domain_thread_func (v=0x7fffffffe070) at domain.c:575
#12 0x00007ffff76196db in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#13 0x00007ffff734271f in clone () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) 
```

`Format` internally uses a global queue.

**Comment by kayceesrk** [comment]
I've introduced a fix for this in #586. But you'd still need to do a bit more work to get the output from multiple domains not interleaving. See https://github.com/ocaml/ocaml/issues/10453#issuecomment-862161428.

**Comment by talex5** [comment]
Thanks! I think it's fine for the client to manage the interleaving.
